### PR TITLE
Add hessian product function to logloss primitve

### PR DESCRIPTION
**Pull request by avolkov-intel**
# Description=
To use logloss objective in newton-cg optimizer it should provide hessian-product callable object.

In this PR I implemented class logloss_hessp<Float> with overloaded operator() to calculate hessian product and implemented a function that calculated raw hessian

### CODE REVIEW

File: cpp/oneapi/dal/backend/primitives/objective_function/BUILD [hunk 1]
```diff
@@ -9,7 +9,8 @@ dal_module(
     auto = True,
     dal_deps = [
         "@onedal//cpp/oneapi/dal/backend/primitives:common",
-        "@onedal//cpp/oneapi/dal/backend/primitives:blas"
+        "@onedal//cpp/oneapi/dal/backend/primitives:blas",
+        "@onedal//cpp/oneapi/dal/backend/primitives:rng"
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/BUILD:13]: Why do you need RNGs here?

Comment (KulikovNikita) [reply cpp/oneapi/dal/backend/primitives/objective_function/BUILD:13]: You can move it to test module instead

File: cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp [hunk 2]
```diff
@@ -21,6 +21,10 @@
 #include "oneapi/dal/test/engine/fixtures.hpp"
 #include "oneapi/dal/table/row_accessor.hpp"
 
+#include "oneapi/dal/backend/primitives/rng/rng_engine.hpp"
+
+// #include "oneapi/dal/backend/primitives/rng/rng_engine.hpp"
+
```
Comment (KulikovNikita) [inline cpp/oneapi/dal/backend/primitives/objective_function/test/logloss_dpc.cpp:27]: Please remove an extra copy

**Review by KulikovNikita — APPROVED** [review]
Please consider looking at my proposals. otherwise no major objections

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss.hpp [hunk 3]
```diff
@@ -88,4 +88,35 @@ sycl::event compute_hessian(sycl::queue& q,
...
+                           ndview<Float, 1>& out,
+                           const event_vector& deps = {});
+
+private:
+    sycl::queue q_;
+    ndarray<Float, 1> raw_hessian_;
+    const ndview<Float, 2> data_;
+    ndarray<Float, 1> buffer_;
+    const Float L2_;
+    const bool fit_intercept_;
```
[the remark this answers is not in the document]
Comment (avolkov-intel) [reply cpp/oneapi/dal/backend/primitives/objective_function/logloss.hpp:117]: Do not htink it is a good idea to increase the number of template parameters for primitive object, that will make class interface more complicated

**Review by KulikovNikita — APPROVED** [review]

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss.hpp [hunk 4]
```diff
@@ -88,4 +88,44 @@ sycl::event compute_hessian(sycl::queue& q,
...
                             const event_vector& deps = {});
 
+template <typename Float>
+sycl::event compute_raw_hessian(sycl::queue& q,
+                                const ndview<Float, 1>& probabilities,
+                                ndview<Float, 1>& out_hessian,
+                                const event_vector& deps = {});
+
+template <typename Float>
+class logloss_hessp {
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss.hpp:98]:
```suggestion
class logloss_hessian {
```

or 
```suggestion
class logloss_hessian_primitive {
```
No need for short and unreadable naming

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp [hunk 5]
```diff
@@ -462,6 +463,120 @@ sycl::event compute_hessian(sycl::queue& q,
...
+    auto* const out_ptr = out.get_mutable_data();
+    ONEDAL_ASSERT(vec.get_dimension(0) == p_ + 1);
+    ONEDAL_ASSERT(out.get_dimension(0) == p_ + 1);
+    auto fill_buffer_event = fill<Float>(q_, buffer_, Float(1), deps);
+    auto out_suf = out.get_slice(1, p_ + 1);
+    auto out_bias = out.get_slice(0, 1);
+    auto vec_suf = vec.get_slice(1, p_ + 1);
+
+    sycl::event fill_out_event = copy(q_, out_suf, vec_suf, deps);
+    sycl::event fill_out0_event = fill<Float>(q_, out_bias, Float(0), deps);
```
Comment (samir-nasibli) [inline cpp/oneapi/dal/backend/primitives/objective_function/logloss_dpc.cpp:527]:
```suggestion
    sycl::event copy_event = copy(q_, out_suf, vec_suf, deps);
    sycl::event fill_out_event = fill<Float>(q_, out_bias, Float(0), deps);
```
Naming should be readable and describable

File: cpp/oneapi/dal/backend/primitives/objective_function/logloss.hpp (hunk 4 above)
Comment (avolkov-intel) [reply cpp/oneapi/dal/backend/primitives/objective_function/logloss.hpp:98]: hessp is shorten name from hessian product

**Pull request merged**